lib/task_manager.js

```diff
@@ -0,0 +1,182 @@
+
+class TaskManager {
+  constructor() {
+    this.queue;
+    this.taskData = {};
+    this.lastPrimaryKey;
+    this.lastExportedBlock;
+    this.consequentTaskIndex = 0;
+  }
+
+  /**
+   * Method for creating a TaskManager instance.
+   * @param {number} maxConcurrentRequests Number of maximum concurrent tasks that should work at the same time
+   * @returns A TaskManager instance
+   */
+  static async create(maxConcurrentRequests) {
+    const tm = new TaskManager();
+    await tm.initQueue(maxConcurrentRequests);
+    return tm;
+  }
+
+  /**
+   * Method for initialization of the queue. It's done in such a way,
+   * because the p-queue package itself does not support CommonJS type importing.
+   * @param {number} maxConcurrentRequests Number of maximum concurrent tasks that should work at the same time
+   */
+  async initQueue(maxConcurrentRequests) {
+    const PQueue = (await import('p-queue')).default;
+    this.queue = new PQueue({ concurrency: maxConcurrentRequests });
+
+    this.queue.on('completed', ([interval, data]) => {
+      this.consequentTaskIndex++;
+      this.#handleNewData(interval, data);
+      this.#maybePauseQueue();
+    });
+  }
+
+  /**
+   * On the completion of a task in the p-queue, the task
+   * should return an array in the form of [interval, events].
+   * These would be set up in the taskData object accordingly in the
+   * correct format
+   * @param {object} interval
+   * @param {Array} newTransformedData 
+   */
+  #handleNewData(interval, newTransformedData) {
+    this.taskData[interval.fromBlock] = { toBlock: interval.toBlock, data: newTransformedData };
+  }
+
+  /**
+   * If the Kafka producer slows down and at the same time we manage
+   * to fetch a lot of data, we'll hit an OOM error at some point, because
+   * of the `taskData` property getting full.
+   * We want to have a border (MAX_TASK_DATA_KEYS), which, if we cross, we'd have the queue
+   * pause the task generation for a bit.
+   */
+  #maybePauseQueue() {
+    if (this.consequentTaskIndex >= MAX_TASK_DATA_KEYS) {
+      if (!this.queue.isPaused) {
+        this.queue.pause();
+        logger.info('Pausing the queue...');
+      }
+    }
+  }
+
+  /**
+   * When the Kafka Producer finishes with the storage, we'd want to check
+   * whether the queue's task generation has been paused, and if positive,
+   * start the queue anew.
+   */
+  restartQueueIfNeeded() {
+    if (this.queue.isPaused) {
+      this.queue.start();
+      this.consequentTaskIndex = 0;
+      logger.info('Resuming the queue...');
+    }
+  }
+
+  /**
+   * Helper method for when we push the `taskData` events' data into the buffer
+   * @param {Array} events Events data that we get from the current key in the `taskData` property
+   * @param {Array} buffer The array that at the end of the primary function would result in the 
+   * combined data, accordingly updated with the primary keys
+   */
+  #updatePrimaryKeysPushToBuffer(events, buffer) {
```

Comment:
No need for this to happen here. The TaskManager can provide the data as it is and it can be decorated on the upper level.